syntax="proto2";
package azino.storage;

option cc_generic_services = true;

// needs to be consistent with the storage/include/status.h
message StorageStatus {
  enum Code {
    Ok = 0;
    NotFound = 1;
    Corruption = 2;
    NotSupported = 3;
    InvalidArgument = 4;
    IOError = 5;
  };
  optional Code error_code = 1 [default = Ok];
  optional string error_message = 2;
}

message PutRequest {
  required string key = 1;
  required string value = 2;
};

message PutResponse {
  optional StorageStatus status = 1;
};

message GetRequest {
  required string key = 1;
};

message GetResponse {
  optional string value = 1;
  optional StorageStatus status = 2;
};

message DeleteRequest {
  required string key = 1;
};

message DeleteResponse {
  optional StorageStatus status = 1;
};

service StorageService {
  rpc Put(PutRequest) returns (PutResponse);
  rpc Get(GetRequest) returns (GetResponse);
  rpc Delete(DeleteRequest) returns (DeleteResponse);
};
